core/src/sumeragi/network_topology.rs

```diff
@@ -125,29 +128,125 @@ impl Topology {
     pub fn leader(&self) -> &PeerId {
         &self.sorted_peers[0]
     }
+
     /// Get proxy tail's peer id.
     pub fn proxy_tail(&self) -> &PeerId {
-        &self.sorted_peers[self.min_votes_for_commit()]
+        &self.sorted_peers[self.min_votes_for_commit() - 1]
     }
+
     /// Add or remove peers from the topology.
     pub fn update_peer_list(&mut self, mut new_peers: HashSet<PeerId>) {
         self.sorted_peers.retain(|peer| new_peers.remove(peer));
         self.sorted_peers.extend(new_peers);
     }
     /// Rotate peers after each failed attempt to create a block.
     pub fn rotate_all(&mut self) {
-        self.sorted_peers.rotate_left(1);
+        self.rotate_all_n(1);
     }
+
+    /// Rotate peers n times where n is a number of failed attempt to create a block.
+    pub fn rotate_all_n(&mut self, n: usize) {
+        let len = self.sorted_peers.len();
+        if let Some(rem) = n.checked_rem(len) {
+            self.sorted_peers.rotate_left(rem);
+        }
+    }
+
     /// Re-arrange the set of peers after each successful block commit.
     pub fn rotate_set_a(&mut self) {
         let rotate_at = self.min_votes_for_commit().min(self.sorted_peers.len());
-        self.sorted_peers[..rotate_at].rotate_left(1);
+        if rotate_at > 0 {
+            self.sorted_peers[..rotate_at].rotate_left(1);
+        }
     }
+
     /// Pull peers up in the topology to the top of the a set while preserving local order.
     pub fn lift_up_peers(&mut self, to_lift_up: &[PublicKey]) {
         self.sorted_peers
             .sort_by_cached_key(|peer| !to_lift_up.contains(&peer.public_key));
     }
+
+    /// Perform sequence of actions after block committed.
+    pub fn update_topology(&mut self, block: &VersionedCommittedBlock, new_peers: HashSet<PeerId>) {
+        self.lift_up_peers(
+            &block
+                .signatures()
+                .map(|s| s.public_key())
+                .cloned()
+                .collect::<Vec<PublicKey>>(),
+        );
+        self.rotate_set_a();
+        self.update_peer_list(new_peers);
+    }
+
+    /// Recreate topology for given block and view change index
+    pub fn recreate_topology(
+        block: &VersionedCommittedBlock,
+        view_change_index: usize,
+        new_peers: HashSet<PeerId>,
+    ) -> Self {
+        let mut topology = Topology::new(block.as_v1().header().committed_with_topology.clone());
+
+        topology.update_topology(block, new_peers);
+
+        // Rotate all once for every view_change
+        topology.rotate_all_n(view_change_index);
+
+        topology
+    }
+
+    /// Check if block's signatures meet requirements for given topology
```

Comment:
Please refer to requirements in the whitepaper, or (better) inline the requirements to the docstring. 

Consensus is important, and good documentation here pays off in hours saved debugging.